jacques/jacques.py

```diff
@@ -0,0 +1,129 @@
+import pandas as pd
+import numpy as np
+import tensorflow as tf
+import abc
+
+class jacques(abc.ABC):
+    def featurize_data(data, target_var="inc_hosp", h=1):
+        """
+        Convert data to tensors containing features x and responses y for each location
+        and time.
+
+        Parameters
+        ----------
+        data: data frame 
+            It has columns location, date, inc_hosp and population
+            and a column (rate) with a response variable to forecast. This data frame needs to be
+            sorted by location and date columns in ascending orer.
+        target_var: string
+            Name of the column in the data frame with the forecast target variable.
+            Default to "inc_hosp"
+        h: integer
+            Forecast horizon. Default to 1
+
+        Returns
+        -------
+        x_train_val: 2D tensor with shape (L*(T-6-1), P = 1) 
+            L is number of lcoation. T is number of time points. P is number of features
+            Each row is a vector x_{l,t} = [x_{l,t,1},...,x_{l,t,P}] of features for some pair 
+            (l, t) in the training set.
+        y_train_val: 1D tensor with length L*(T-6-1)
+            Each value is a forecast target variable value in the training set.
+            y_{l, t} = z_{l, 1, t+h}
+        x_T: 2D tensor with shape (L, P = 1)
+            Each value is test set feature for each location at forecast date.
+        """
+        assert target_var in data.columns
+
+        # data has columns date, location, inc hosp, rate and population
+        T = max(data['date'])
+
+        # create a column for 7-day moving avg of rates
+        # for each location, the first 6 days have nan
+        data['moving_avg_rate'] = data.groupby('location').rolling(7)['rate'].mean().reset_index(drop=True)
+        
+        # list of indices where rates are nan
+        na_idx = data.loc[pd.isna(data["moving_avg_rate"]), :].index
+        
+        # take out nans in data
+        tmp = data.drop(na_idx)
+        
+        # list of indices where date is the forecast date in tmp
+        T_idx = tmp.loc[tmp["date"]== T,:].index
+
+        # create x_T using data with date = forecast_date (T)
+        data_T = data.loc[data["date"]== T,:]
+    
+        # x_T is (L, P = 1)
+        x_T = data_T['moving_avg_rate'].values.reshape(-1, 1)
+        
+        # x_train_val is (L*(T-6-1), P = 1) 
+        train_val = tmp.drop(T_idx)
+        x_train_val = train_val['moving_avg_rate'].values.reshape(-1, 1)
+
+        # tmp has value on forecast_date for each location
+        grp = tmp[target_var].groupby(tmp['location'])
+        # y_train_val is (L*(T-6-1), 1) 
+        # pad h-1 numbers of nan at the end of each v means extending the array for every date before forecast_date (T)
+        # -1 is for forecast_date (T)
+        y_train_val = np.vstack([np.pad(v, (0, h-1), constant_values=np.nan)[h:].reshape(-1, 1) for _, v in grp])
```

Comment:
I trust that you have got this right, but I'd suggest that something like the following would be easier for us to parse:
```suggestion
        y_train_val = data.groupby('location')[target_var].shift(-h)
```

If we do this, we will have to clean up missing values later on.